brandon_tan/partner_selector.py

```diff
@@ -0,0 +1,483 @@
+
+            if total_euclidean_distance < lowest_euclidean_distance:
+                lowest_euclidean_distance = total_euclidean_distance
+                best_Q = partners
+
+            partner_data[i] = {
+                'Grouping': partners,
+                'Total Euclidean Distance': total_euclidean_distance
+            }
+
+        self.partner_data = partner_data
+        self.best_Q = best_Q
+
+        return best_Q
+
+    def _diagonal_distance(self, vector: np.ndarray):
+        """
+        Helper private function to calculate the euclidean distance between a given point
+        and the hyper diagonal of the space.
+
+        The distance is computed by taking a perpendicular point on the hyper diagonal
+        as v1 = (k,k,k,k). Our point of interest is v2 = (w,y,x,z). Solving for the first derivative
+        of the distance d(v1,v2), we obtain a formula to compute the distance.
+
+        :param vector: The coordinates of the point.
+        :type vector: np.ndarray
+        :return: Euclidean distance between point and hyper diagonal of space.
+        :rtype: np.float
+        """
+        # Using a loop so that if the user chooses higher dimensions,
+        # this will still work, rather than a hardcoded formula.
+        euclidean_distance = 0
+        for i in vector:
+            # Math can be found in docstring.
+            mask = vector != i
+            other_terms_sum = sum(vector[mask])
+            euclidean_distance += ((3 * i - other_terms_sum) / 4) ** 2
+        return euclidean_distance
+
+    def print_info(self) -> pd.DataFrame:
+        """
+        Prints descriptive statistics of the selection process.
+        This is inspired by pd.describe().
+
+        :raises Exception: If selection process has not been done, an error will be raised.
+        :return: Dataframe containing descriptive statistics of the selection process.
+        :rtype: pd.DataFrame
+        """
+        if self.partner_data is None:
+            raise Exception("Potential parters have not been identified yet. Please run select() before this method.")
+
+        info_df = pd.DataFrame.from_dict(self.partner_data).transpose()
+        info_df = info_df[info_df.columns[1]]
+        info_df = info_df.astype(float)
+        info_df = pd.DataFrame(info_df.describe())
+        return info_df
+
+    def plot_selected_partners(self):
+        """
+        Plots the historical log prices of the best partners that was identified
+        by the method of choice by the user.
+
+        :raises Exception: If selection has not been done, error will be raised.
+        """
+        if self.best_Q is None:
+            raise Exception("The best parters have not been identified yet. Please run select() before this method.")
+
+        fig, ax = plt.subplots(nrows=1, ncols=1, figsize=(15, 5))
+
+        for security in self.best_Q:
+            rets = np.log(self.price_data[security]).pct_change()
+            rets = np.cumprod(1+rets)
```

Comment:
No. You don't take the cumprod of log returns, they are additive. You take the cumsum.

Reply:
These are serious mistakes.

Reply:
This is a careless mistake on my part. I apologise for it.

The error arose as the visual graph passed the eyeball test, so I assumed that it was correct.